[Merged by Bors] - feat(ring_theory/noetherian): maximal among set iff Noetherian

knaka3435 opened the pull request:
Main theorem is `set_has_maximal_iff_noetherian,` which relates well foundedness of `<` to being noetherian.
Most notably a result of
`well_founded.well_founded_iff_has_max'` provides the fact that on a partial ordering, `well_founded >` is equivalent to each nonempty set having a maximal element.
`well_founded.well_founded_iff_has_min` provides an analogous fact for `well_founded <`.

Some other miscellaneous lemmas are as follows
`localization_map.integral_domain_of_local_at_prime` is the localization of an integral domain at a prime's complement is an integral domain
`ideal.mul_eq_bot` is the fact that in an integral domain if I*J = 0, then at least one is 0.
`submodule.nonzero_mem_of_gt_bot` is that if ⊥ < J, then J has a nonzero member.
`lt_add_iff_not_mem` is that b is not a member of J iff J < J+(b).
`bot_prime` states that 0 is a prime ideal of an integral domain.

---
<!-- put comments you want to keep out of the PR commit here -->

golfing/renaming/rehoming/suggestions are all welcome!

jalex-stark committed:
init

knaka3435 committed:
placing, very buggy(?)

knaka3435 committed:
move

knaka3435 committed:
fix(?)

knaka3435 committed:
fixing bugs

knaka3435 committed:
dup instance

knaka3435 committed:
rearranging

knaka3435 committed:
golf

knaka3435 committed:
cleanup

awainverse commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -138,6 +138,10 @@ end
 
+-- probably redundant
+lemma lt_add_nonmem (I : ideal α) (a ∉ I) : I < I + ideal.span{a} :=
+by {exact submodule.lt_add_nonmem I a H,}
```
This could be in term mode: 
```suggestion
submodule.lt_add_nonmem I a H
```

awainverse replied:
But also you could skip it, as it isn't needed for the following to work:
```
example (I : ideal α) (H : a ∉ I) : I < I + ideal.span {a} := I.lt_add_nonmem a H
```

awainverse commented on src/ring_theory/ideal/operations.lean:
```diff
@@ -859,6 +859,27 @@ end
 
 end ideal
 
+lemma ideal_mul_eq_zero (R : Type* )[integral_domain R] {I J : ideal R} : (I * J = ⊥) ↔ I = ⊥ ∨ J = ⊥ :=
```
```suggestion
lemma ideal_mul_eq_bot (R : Type*) [integral_domain R] {I J : ideal R} : (I * J = ⊥) ↔ I = ⊥ ∨ J = ⊥ :=
```
I think it's standard in mathlib to refer to this ideal as `bot` rather than `zero`. Also a bit of spacing.

awainverse commented on src/ring_theory/noetherian.lean:
```diff
@@ -360,6 +360,55 @@ begin
+open_locale classical
+/-- A module is Noetherian iff every nonempty set of submodules has a maximal submodule among them
+-/
+theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] : (∀(a : set $ submodule R M), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I=M') ↔ is_noetherian R M :=
```
Can this be split into a statement about well-foundedness in partial orders, chained with `is_noetherian_iff_well_founded`?

knaka3435 replied:
To clarify, you want something of the form `well_founded r` implies the first definition [given here](https://en.wikipedia.org/wiki/Well-founded_relation)? Ah, `well_founded.has_min` works.

awainverse commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -474,6 +478,20 @@ end pi
 end ideal
 
+--Move this somewhere better
+lemma zero_prime [integral_domain α] : (⊥ : ideal α).is_prime :=
```
```suggestion
lemma bot_prime [integral_domain α] : (⊥ : ideal α).is_prime :=
```

knaka3435 committed:
Apply suggestions from code review

Co-authored-by: Aaron Anderson <65780815+awainverse@users.noreply.github.com>

knaka3435 committed:
fix

knaka3435 committed:
Mergefix

knaka3435 committed:
golf

knaka3435 committed:
golf

awainverse commented on src/ring_theory/noetherian.lean:
```diff
@@ -362,7 +362,7 @@ end
 /-- A module is Noetherian iff every nonempty set of submodules has a maximal submodule among them
 -/
-theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] : (∀(a : set $ submodule R M), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I=M') ↔ is_noetherian R M :=
+theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] : (∀(a : set $ submodule R M), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I = M') ↔ is_noetherian R M :=
```
It looks like most of your proof doesn't rely on anything about `submodule`s beyond the fact that they're a partial order...
I think you could break off most of your proof into a separate lemma, about partial orders, perhaps show that if `P` is a partial order, then > is `well_founded` iff 
```
(∀(a : set P), a.nonempty → ∃ (M' ∈ a), ∀ (I ∈ a), M' ≤ I → I = M')
```

ChrisHughes24 replied:
Something very similar should be true for well founded relations. I proved the right direction quite easily 
```
exact λ wf a ha, ⟨well_founded.min wf a ha, well_founded.min_mem _ _ _, 
      λ I hI hminI, (lt_or_eq_of_le hminI).elim 
        (λ h, (well_founded.not_lt_min wf _ _ hI h).elim) 
        eq.symm⟩
```
I could find a proof for the left direction in mathlib, but this would be useful

knaka3435 replied:
I would agree, but for now I added your code under `well_founded.has_min'` in `rel_classes`, and it has made the back half significantly shorter :+1:

Vierkantor replied:
Here is a proof for the converse direction:
```lean
lemma well_founded_of_exists_max {α : Type*} [partial_order α]
  (exists_max : ∀ (p : set α), p.nonempty → ∃ m ∈ p, ∀ x ∈ p, m ≤ x → x = m) :
  well_founded ((>) : α → α → Prop) :=
⟨begin
  set counterexamples := {x : α | ¬ acc gt x},
  intro x,
  by_contra hx,
  obtain ⟨m, m_mem, hm⟩ := exists_max counterexamples ⟨x, hx⟩,
  refine m_mem (acc.intro _ (λ y y_gt_m, _)),
  by_contra hy,
  exact ne_of_lt y_gt_m (hm y hy (le_of_lt y_gt_m)).symm
end⟩
```
Could you add that to `order.rel_classes` and use it here?

knaka3435 committed:
has_min'

knaka3435 committed:
formatting

Vierkantor commented on src/linear_algebra/basic.lean:
```diff
@@ -436,6 +436,11 @@ variables (R)
+lemma nonzero_mem_of_gt_bot {I : submodule R M} (gt_bot : ⊥ < I) : ∃ a : I, a ≠ 0 :=
+begin
+  have h := (submodule.lt_iff_le_and_exists.1 gt_bot).2, tidy,
+end
```
```suggestion
lemma nonzero_mem_of_bot_lt {I : submodule R M} (bot_lt : ⊥ < I) : ∃ a : I, a ≠ 0 :=
begin
  have h := (submodule.lt_iff_le_and_exists.1 bot_lt).2,
  tidy,
end
```

Vierkantor commented on src/linear_algebra/basic.lean:
```diff
@@ -845,6 +850,17 @@ lemma span_singleton_le_iff_mem (m : M) (p : submodule R M) :
+  have h3 := h1 h2 ile,
+  have : a ∈ submodule.span R {a}, exact submodule.mem_span_singleton_self a,
+  tidy,
+end
```
You should be able to use `lt_iff_le_and_exists` to shorten the proof.

knaka3435 replied:
Thanks for the hint! The proof is a lot more straightforward now.

Vierkantor commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,13 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+by  { exact λ wf a ha, ⟨well_founded.min wf a ha, well_founded.min_mem _ _ _,
+      λ I hI hminI, (lt_or_eq_of_le hminI).elim
+        (λ h, (well_founded.not_lt_min wf _ _ hI h).elim)
+        eq.symm⟩ }
```
I golfed this proof. Steps:

 * `by { exact e }` is just `e`
 * If your proof starts with `λ wf`, you can make `wf` a named parameter.
 * `or.elim (λ h, false.elim _)` is `or.resolve_left`
 * update variable names to match conventions in this section

```suggestion
theorem has_min' [partial_order α] (wf : well_founded (gt : α → α → Prop))
  (p : set α) (hp : p.nonempty) : ∃ m ∈ p, ∀ x ∈ p, m ≤ x → x = m :=
⟨min wf p hp,
 min_mem _ _ _,
 λ I hI hminI, eq.symm ((lt_or_eq_of_le hminI).resolve_left (not_lt_min wf _ _ hI))⟩
```

(Not something that you should fix, but I could not find the characterization of the minimum in a partial order `eq_min_iff_le_min : x = well.founded.min p ↔ x ∈ p ∧ x ≤ well.founded.min p` and/or `min_unique : x = well.founded.min p ↔ x ∈ p ∧ ∀ y ∈ s, x ≤ y`).

Vierkantor commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -474,6 +474,20 @@ end pi
 
 end ideal
 
+--Move this somewhere better
```
This can go after `span_singleton_prime` in this file, line 168.
```

knaka3435 replied:
It seems to be pretty buggy if I do that. I tried to fix it by using `@one_ne_zero α _ _`, but I ended up with `h : 1 = 0` and `⊢ 1 = 0` all as elements of `α`. I think the problem was that `0` in the goal ended up being a bit different in `h`, `add_monoid.to_has_zero` and `mul_zero_class.to_has_zero`, and generally, it seems the type inferencer is not happy with either line. Oddly enough, it's fine where it is, but moving it up inside ideal just breaks it (although this might just be my inexperience showing).

Vierkantor replied:
The problem was an instance conflict: the type `α` already had a declared `comm_ring` instance, so adding an `integral_domain` instance as well will give it two incompatible instances of `0`, `1`, `+`, `*`, ... I took the liberty of pushing a commit fixing this.

Vierkantor commented on src/ring_theory/ideal/basic.lean:
```diff
@@ -474,6 +474,20 @@ end pi
+    have x_or_y0 : x = 0 ∨ y = 0,
+    exact zero_eq_mul.mp (eq.symm h1),
+    tauto },
+end
```
I also golfed the proof here. Steps:

 * `have h1 : a ↔ b := e, rw h1` is just `rw e`
 * I know that `x ∈ ⊥ ↔ x = 0`, this is called `submodule.mem_bot`, so rewrite with that everywhere
 * Now the first goal is `¬ 1 = 0`, which is called `one_ne_zero`. So `intros, apply one_ne_zero, rw [...]` works to solve the first goal.
 * Now the second goal is `x * y = 0 ⊢ x = 0 ∨ y = 0`, which is called `mul_eq_zero`, so `intros, apply mul_eq_zero, simp only [...]` works.
 * Convert the short tactical proofs to term-mode proofs.

```suggestion
⟨λ h, one_ne_zero (by rwa [ideal.eq_top_iff_one, submodule.mem_bot] at h),
 λ x y h, mul_eq_zero.mp (by simpa only [submodule.mem_bot] using h)⟩
```

Vierkantor commented on src/ring_theory/ideal/operations.lean:
```diff
@@ -859,6 +859,27 @@ end
 
 end ideal
 
+lemma ideal_mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} : (I * J = ⊥) ↔ I = ⊥ ∨ J = ⊥ :=
```
This can go after `pow_le_pow`, line 340. Then we can drop the `ideal` from the name:
```suggestion
lemma mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} : (I * J = ⊥) ↔ I = J ∨ J = ⊥ :=
```

Vierkantor commented on src/ring_theory/ideal/operations.lean:
```diff
@@ -859,6 +859,27 @@ end
+  specialize hij (i * j'),
+  cases eq_zero_or_eq_zero_of_mul_eq_zero (hij $ ideal.mul_mem_mul hi hj),
+  assumption, simpa only [],
+end
```
Golfed this proof. Steps:

 * `j` was not used (it will be equal to `0` anyway), so we can get rid of those lines.
 * We don't need to specify `J` in `rw ideal.mul_bot J`, so the two `rw` lines become the same and we can write `cases a; rw`.
 * We can shorten the case distinction on whether `J = ⊥` with the following trick: `refine or_iff_not_imp_left.mpr (λ J_ne_bot, _),`
 * We rewrite `submodule.eq_bot_iff` at the goal and at `hij`, so we can replace it with one `rw at ⊢ hij`.
 * The `cases` where one case is impossible can be replaced with `or.resolve_right`
 * Translate the tactical proof into a term-mode proof dropping into tactic mode when needed

```suggestion
lemma ideal_mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} :
  I * J = ⊥ ↔ I = ⊥ ∨ J = ⊥ :=
⟨λ hij, or_iff_not_imp_left.mpr (λ I_ne_bot, J.eq_bot_iff.mpr (λ j hj,
  let ⟨i, hi, ne0⟩ := I.ne_bot_iff.mp I_ne_bot in
    or.resolve_left (mul_eq_zero.mp ((I * J).eq_bot_iff.mp hij _ (mul_mem_mul hi hj))) ne0)),
 λ h, by cases h; rw [← ideal.mul_bot, h, ideal.mul_comm]⟩
```

Vierkantor commented on src/ring_theory/localization.lean:
```diff
@@ -954,6 +954,18 @@ def integral_domain_localization {M : submonoid A} (hM : M ≤ non_zero_divisors
+/--
+The localization of an integral domain at the complement of a prime ideal is an integral domain.
+-/
+instance integral_domain_of_local_at_prime {P : ideal A} (hp_prime : P.is_prime) :
```
I'm not an expert on typeclasses, but I think the rule of thumb is that all parameters of an instance should be either `{implicit}` or `[typeclasses]`. And `ideal.is_prime` is a typeclass, so it should be:
```suggestion
instance integral_domain_of_local_at_prime {P : ideal A} [P.is_prime] :
```

knaka3435 replied:
I'm not sure either, but I think that it might be cumbersome to use instances if you want to consider multiple localizations of the same ring (although I'm not sure if there are enough interesting uses to warrant this).
EDIT: I'm going to change it back to explicit because based off some testing, I the typeclass inferencer can't really figure it out.

Vierkantor replied:
Aha, it looks like the `is_prime` instance is already found by unification with `localization.at_prime`. So it should indeed work fine like you had originally.

Vierkantor commented on src/ring_theory/localization.lean:
```diff
@@ -954,6 +954,18 @@ def integral_domain_localization {M : submonoid A} (hM : M ≤ non_zero_divisors
+  { have := ideal.zero_mem P, simpa },
+  have : P.prime_compl ≤ non_zero_divisors A, exact le_non_zero_divisors_of_domain zero_non_mem,
+  exact integral_domain_localization this,
+end
```
This can be a term-mode proof:
```suggestion
integral_domain_localization (le_non_zero_divisors_of_domain (by simpa only [] using P.zero_mem))
```

Vierkantor commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,13 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
   (p : set α) (h : p.nonempty) {x} (xp : x ∈ p) : ¬ r x (H.min p h) :=
 let ⟨_, h'⟩ := classical.some_spec (H.has_min p h) in h' _ xp
 
+theorem has_min' [partial_order α] : (well_founded (gt : α → α → Prop)) →
```
This definition is should be called `has_max` because we get the largest element of the set `a` instead of the smallest. I think we should have both `has_min` (assuming `(<)` is well-founded) and `has_max` (assuming `(>)` is well-founded.

knaka3435 committed:
review suggestions

Co-authored-by: Vierkantor <Vierkantor@users.noreply.github.com>

knaka3435 committed:
well_ordered_iff_

knaka3435 committed:
fix

Vierkantor committed:
Move `ideal.bot_prime` to the other lemmas about prime ideals

To make the move work, we have to make a new type `R` which carries the
`integral_domain` instance, because it would otherwise conflict with the
`comm_ring` instance on `α` (see also `span_singleton_eq_span_singleton`).

Vierkantor commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,46 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+  { intros wf p hp,
+    exact ⟨min wf p hp,
+    min_mem _ _ _,
+    λ I hI hminI, eq.symm ((lt_or_eq_of_le hminI).resolve_left (not_lt_min wf _ _ hI))⟩ },
```
```suggestion
    exact ⟨min wf p hp,
           min_mem _ _ _,
           λ I hI hminI, eq.symm ((lt_or_eq_of_le hminI).resolve_left (not_lt_min wf _ _ hI))⟩ },
```

Vierkantor commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,46 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+  { intros wf p hp,
+    exact ⟨min wf p hp,
+    min_mem _ _ _,
+    λ I hI hminI, ((lt_or_eq_of_le hminI).resolve_left (not_lt_min wf _ _ hI))⟩ },
```
```suggestion
           min_mem _ _ _,
           λ I hI hminI, ((lt_or_eq_of_le hminI).resolve_left (not_lt_min wf _ _ hI))⟩ },
```

Vierkantor commented on src/ring_theory/noetherian.lean:
```diff
@@ -358,6 +358,40 @@ begin
+    rw is_noetherian_iff_well_founded at h,
+    exact well_founded.well_founded_iff_has_max.1 h A ha },
+end
+
```
With the new definition, we can shorten this to:
```suggestion
/-- A module is Noetherian iff every nonempty set of submodules has a maximal submodule among them. -/
theorem set_has_maximal_iff_noetherian {R M} [ring R] [add_comm_group M] [module R M] :
  (∀ a : set $ submodule R M, a.nonempty → ∃ M' ∈ a, ∀ I ∈ a, M' ≤ I → I = M') ↔ is_noetherian R M :=
by rw [is_noetherian_iff_well_founded, well_founded.well_founded_iff_has_max]
```

Vierkantor reviewed (changes requested):
Just a few smaller comments left, then I think everything looks good!

knaka3435 committed:
Apply suggestions from code review

Co-authored-by: Vierkantor <Vierkantor@users.noreply.github.com>

Vierkantor reviewed (approved):
Looks good to me!

knaka3435 committed:
lt_add_nonmem iff

ChrisHughes24 commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,46 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+
+theorem well_founded_iff_has_min [partial_order α] : (well_founded (has_lt.lt : α → α → Prop)) ↔
+  ∀ (p : set α), p.nonempty → ∃ m ∈ p, ∀ x ∈ p, x ≤ m → x = m :=
+begin
```
the prood `@well_founded_iff_has_max (order_dual α) _` should work here.

knaka3435 replied:
ah, thanks for that!

ChrisHughes24 commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,46 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
   (p : set α) (h : p.nonempty) {x} (xp : x ∈ p) : ¬ r x (H.min p h) :=
 let ⟨_, h'⟩ := classical.some_spec (H.has_min p h) in h' _ xp
 
+theorem well_founded_iff_has_max [partial_order α] : (well_founded ((>) : α → α → Prop) ↔
```
Can this be generalised to an arbitrary relation `r` instead of `>`? The condition `m ≤ x → x = m` could be replaced with `¬ r x m`.

ChrisHughes24 commented on src/linear_algebra/basic.lean:
```diff
@@ -845,6 +851,25 @@ lemma span_singleton_le_iff_mem (m : M) (p : submodule R M) :
   span R {m} ≤ p ↔ m ∈ p :=
 by rw [span_le, singleton_subset_iff, mem_coe]
 
+lemma lt_add_nonmem {I : submodule R M} {a : M} : I < I + submodule.span R {a} ↔ a ∉ I :=
```
```suggestion
lemma lt_add_iff_not_mem {I : submodule R M} {a : M} : I < I + span R {a} ↔ a ∉ I :=
```

ChrisHughes24 commented on src/ring_theory/ideal/operations.lean:
```diff
@@ -339,6 +339,13 @@ begin
   exact le_trans (mul_le_inf) (inf_le_left)
 end
 
+lemma ideal_mul_eq_bot {R : Type*} [integral_domain R] {I J : ideal R} :
```
Is this in the` ideal` namespace? It could be called `ideal.mul_eq_bot`

knaka3435 committed:
review

bryangingechen commented on src/linear_algebra/basic.lean:
```diff
@@ -845,6 +851,25 @@ lemma span_singleton_le_iff_mem (m : M) (p : submodule R M) :
+  use a,
+  split, swap, assumption,
+  have : span R {a} ≤ I + span R{a}, exact le_sup_right,
+  exact this (@mem_span_singleton_self R _ _ _ _ a), }
```
```suggestion
    apply lt_iff_le_and_exists.mpr, split,
    simp only [add_eq_sup, le_sup_left],
    use a,
    split, swap, { assumption, },
    { have : span R {a} ≤ I + span R{a} := le_sup_right,
      exact this (@mem_span_singleton_self R _ _ _ _ a), },
```

bryangingechen commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,43 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+  obtain ⟨m, m_mem, hm⟩ := exists_max counterexamples ⟨x, hx⟩,
+  refine m_mem (acc.intro _ ( λ y y_gt_m, _)),
+  by_contra hy,
+  exact hm y hy y_gt_m,
```
```suggestion
  { exact has_min, },
  { set counterexamples := { x : α | ¬ acc r x},
    intro exists_max,
    fconstructor,
    intro x,
    by_contra hx,
    obtain ⟨m, m_mem, hm⟩ := exists_max counterexamples ⟨x, hx⟩,
    refine m_mem (acc.intro _ ( λ y y_gt_m, _)),
    by_contra hy,
    exact hm y hy y_gt_m, },
```

bryangingechen commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,43 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+    exact lt_irrefl x nge },
+  { intros ngt xle,
+  contrapose! ngt,
+  exact lt_of_le_of_ne xle (ne.symm ngt) }
```
```suggestion
    contrapose! ngt,
    exact lt_of_le_of_ne xle (ne.symm ngt) }
```

bryangingechen commented on src/linear_algebra/basic.lean:
```diff
@@ -845,6 +851,25 @@ lemma span_singleton_le_iff_mem (m : M) (p : submodule R M) :
+    by_contra akey,
+    have h1 : I + span R {a} ≤ I,
+    { simp only [add_eq_sup, sup_le_iff],
+    split, exact le_refl I, exact (span_singleton_le_iff_mem a I).mpr akey, },
```
```suggestion
      split,
      { exact le_refl I, },
      { exact (span_singleton_le_iff_mem a I).mpr akey, } },
```

bryangingechen reviewed (commented):
Just some minor indentation changes.

knaka3435 committed:
spacing

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

knaka3435 committed:
fix

knaka3435 committed:
fix

ChrisHughes24 commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,43 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
   (p : set α) (h : p.nonempty) {x} (xp : x ∈ p) : ¬ r x (H.min p h) :=
 let ⟨_, h'⟩ := classical.some_spec (H.has_min p h) in h' _ xp
 
+theorem well_founded_iff_has_max  {α} {r : α → α → Prop} : (well_founded r) ↔
```
```suggestion
theorem well_founded_iff_has_min  {α} {r : α → α → Prop} : (well_founded r) ↔
```

ChrisHughes24 commented on src/order/rel_classes.lean:
```diff
@@ -301,6 +301,43 @@ theorem not_lt_min {α} {r : α → α → Prop} (H : well_founded r)
+    exact hm y hy y_gt_m, },
+end
+
+lemma le_imp_eq_iff_not_gt {α} [partial_order α] {x y : α} : x ≤ y → y = x ↔ ¬ x < y :=
```
```suggestion
lemma eq_iff_not_lt_of_le {α} [partial_order α] {x y : α} : x ≤ y → y = x ↔ ¬ x < y :=
```

ChrisHughes24 reviewed (changes requested):
Minor name changes. You can merge after the changes are made. bors d+

knaka3435 committed:
renaming

Co-authored-by: Chris Hughes <33847686+ChrisHughes24@users.noreply.github.com>

knaka3435 committed:
fix
